### DISCUSSION TITLE
Including a C++ library into the setup of a New Application Program Project

### DISCUSSION THREAD

**Discussion by Sam-Vervaeck** [stnolting/neorv32#963] [General]
The TLDR of this question is to ask about the c++ compilation version being used by neorv32 and if it can be controlled.

I want to try to run a program using the following library on the neorv_32: [SEAL](https://github.com/microsoft/SEAL). Now, this library does not support riscv natively so I have been hacking a bit around in the files to see how far I can go and maybe it will end up running. (I want to see how big the performance gap is between running it on Intel silicon vs riscv embedded silicon).

I have gotten already pretty far by turning a lot of special third library optimizations off. The current errors that I am facing all complain that functions such as "round" and "log2" are not a member of "std". Online [Q&A](https://stackoverflow.com/questions/12696764/round-is-not-a-member-of-std#12696794) says this is because the C++ version being used in compilation is not modern enough.

I had a look at the common.mk file to see what compiler settings are being used. I don't see any specific C++ flags however. The NEO_CXXFLAGS variable makes use of the CXXFLAGS variable but that one is not defined? I ran "make -p" as well and on my linux machine the CXXFLAGS also has no default so it is truly empty. 

To end my exposition and give my detailed questions now: 
What version does the neorv32 compiler use for C++? Is this something that is dependent on the compiler toolchain i have installed on my machine. If so, is there an easy command I can run to set this to a more modern version such as C++11? Googling this stuff about riscv toolchains doesn't really give clear answers.

**Comment by Sam-Vervaeck** [comment]
Also, if this is very very complicated stuff I will simply drop this research question and move on to something else. It is not that crucial to have the library running, but would have been interesting.

**Comment by stnolting** [comment]
Hey @Sam-Vervaeck.

> The TLDR of this question is to ask about the c++ compilation version being used by neorv32 and if it can be controlled.

This is a hard question. C++ support is still experimental (at least from my side) as I am not very familiar with that.

First of all, you'll need to change (or better: override) the compiler to use "`g++`":

https://github.com/stnolting/neorv32/blob/00d248cb1933c727d4fdbe4c1f8a0749d7e85b74/sw/common/common.mk#L101

We already have a minimal C++ example here: https://github.com/stnolting/neorv32/blob/main/sw/example/hello_cpp/main.cpp

When I put this code in there

```c
  double level = 3.231;
  volatile double res = 0.02 * round(level);
```

and also include math.h then I can successfully compile everything. However, I am not sure if this helps...

Anyway, we should improve the C++ support / example base. :+1:

**Reply by Sam-Vervaeck** [reply]
> This is a hard question. C++ support is still experimental (at least from my side) as I am not very familiar with that.
> First of all, you'll need to change (or better: override) the compiler to use "g++":

Ah, I had concluded from the second answer to this [post](https://stackoverflow.com/questions/172587/what-is-the-difference-between-g-and-gcc#172592) that I could skip this part, my bad. I'll look into your suggestion a bit tomorrow and see if I get it up and running. Thanks for the input already!

**Reply by Sam-Vervaeck** [reply]
> First of all, you'll need to change (or better: override) the compiler to use "g++"

I did this, but it does not seem to change the errors the SEAL library is throwing me when i include it. I also did a manual on the the toolchain I am using to check what C++ version it uses as default. [This](https://stackoverflow.com/questions/44734397/which-c-standard-is-the-default-when-compiling-with-g#44735016) returned `#define __cplusplus 201703L` so that seems up to date at least to me:

`riscv32-unknown-elf-g++ -dM -E -x c++  /dev/null | grep -F __cplusplus`

I'm going to leave this here for now. Who knows somebody gets it working in the future :). If you prefer, you can mark this as an answer to this thread or leave it unanswered for now.

**Reply by stnolting** [reply]
The library seems to be plain C++ code, so it should compile also for RISC-V. 🤔

**Reply by Sam-Vervaeck** [reply]
If you change to g++ it still complains in some external libraries it makes use of and in the system check done in this file:
https://github.com/microsoft/SEAL/blob/119dc32e135cb89c1062076a69310d4413ebc824/native/src/seal/util/defines.h#L78-L85

**Reply by stnolting** [reply]
Oh, seems like it requires an actual (desktop) operating system... 🙈
